### DISCUSSION TITLE
How to set disk IOPS limit using govc?

### DISCUSSION THREAD

**Discussion by Heiko-san** [vmware/govmomi#2801] [Community Help] [answered]
Hi,

What I want to achieve is to set the IOPS limit for the disk of an existing vm.
I mean the setting that goes into vm's config as "sched.scsi0:0.throughputCap=xy", which you set in GUI by "edit settings" -> go to the disk -> "Limit IOPs"  -> if you switch to "user defined" you can enter any number ...

- I found `govc vm.disk.change` but it seems I can only set disk size there.
- I see the disk using `govc device.info -vm myvm disk-1000-0`
- There is also a `govc disk.ls`, but it doesn't show anything

How can I set this limit?

Update:
- I found this, but doesn't seem to work either (even if it doesn't show any errors): `govc vm.change -vm myvm -e sched.scsi0:0.throughputCap=123`

**Comment by embano1** [comment]
Is `extraConfig` what you're looking for? https://github.com/vmware/govmomi/blob/master/govc/USAGE.md#vmchange

**Comment by Heiko-san** [comment]
I actually tried this (see above, the "-e" switch).

**Reply by embano1** [reply]
Sorry, my bad :( Can you confirm that setting the extended attribute changes the vmx or doesn't it do anything at all?

**Reply by Heiko-san** [reply]
It seems to do nothing at all, the parameter I tried is the one that appears in the vmx file, if I configure the limit in the GUI.
But setting it with "-e" neither sets it in vmx file, nor is the limit shown if I edit the VM in the GUI.

On the other hand: If I download the vmx file, set the parameter in the file, upload it again and restart the VM ... The setting is still not reflected if I edit the VM in GUI.
So maybe it's not a govmomi issue, maybe it's just the wrong way of doing this. (? I'm not that familiar with the vmware API)

**Reply by embano1** [reply]
I think I found the reason. The `extraConfig` way might not be used anymore (though the UI seems to set the value for backwards compat?). Anyways, I found that the API to use is [this](https://vdc-download.vmware.com/vmwb-repository/dcr-public/8946c1b6-2861-4c12-a45f-f14ae0d3b1b9/a5b8094c-c222-4307-9399-3b606a04af55/vim.StorageResourceManager.IOAllocationInfo.html) one.

If you look at the VM configuration you'll find this configuration:

```json5
"StorageIOAllocation": {
              // -1 is unlimited, respectively the value you specify via API/UI
              "Limit": -1,
              "Shares": {
                "Shares": 1000,
                "Level": "normal"
              },
              "Reservation": 0
            }
```

Currently `govc` does not allow to specify this property so the only way is to use the `govmomi` library currently. See https://github.com/vmware/govmomi/issues/396 for related discussion.

**Comment by embano1** [comment, marked as answer]
I created a draft PR for `govc`: https://github.com/vmware/govmomi/pull/2807

**Reply by Heiko-san** [reply]
cool, thanks a lot :)

**Reply by embano1** [reply]
Np, can you please mark this discussion as answered then so it‘s resolved?

**Reply by Heiko-san** [reply]
Stupid question: How? :P I can't find a button or something to resolve it/ mark it as answered. :|

**Reply by embano1** [reply]
Alright, I was wondering the same (and thought only you could do it as author). Let me lock the conversation then. Perhaps a bug in Github.

**Reply by tenthirtyam** [reply]
Moved to Q&A and marked as answered.
